File: autozig-math/src/rect.rs
use autozig::include_zig;
use crate::Vec2;

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Rect {
    pub min: Vec2,
    pub max: Vec2,
}

include_zig!("zig/rect.zig", {
    fn rect_new(min: Vec2, max: Vec2) -> Rect;
    fn rect_width(self_: Rect) -> f32;
    fn rect_height(self_: Rect) -> f32;
});

impl Rect {
    pub fn new(min: Vec2, max: Vec2) -> Self {
        rect_new(min, max)
    }

    pub fn width(&self) -> f32 {
        rect_width(*self)
    }

    pub fn height(&self) -> f32 {
        rect_height(*self)
    }
}

File: autozig-math/zig/rect.zig
const std = @import("std");
const Vec2 = @import("vec2.zig").Vec2;
// Vec2 is available globally via autozig merge

pub const Rect = extern struct {
    min: Vec2,
    max: Vec2,

    pub fn new(min: Vec2, max: Vec2) Rect {
        return .{
            .min = min,
            .max = max,
        };
    }

    pub fn width(self: Rect) f32 {
        return self.max.x - self.min.x;
    }

    pub fn height(self: Rect) f32 {
        return self.max.y - self.min.y;
    }
};

export fn rect_new(min: Vec2, max: Vec2) Rect {
    return Rect.new(min, max);
}

export fn rect_width(self: Rect) f32 {
    return self.width();
}

export fn rect_height(self: Rect) f32 {
    return self.height();
}
